### DISCUSSION TITLE
CocoaPods fails with `current_thread_backend': uninitialized constant ActiveSupport::XmlMini::IsolatedExecutionState (NameError)

### DISCUSSION THREAD

**Discussion by vysotsky** [codemagic-ci-cd/codemagic-docs#1842] [Q&A]
### If you are looking for help, please confirm the following...

- [X] I confirm I have searched the [Docs](https://docs.codemagic.io), [Codemagic Sample Projects](https://github.com/codemagic-ci-cd/codemagic-sample-projects), and GitHub [Discussions](https://github.com/orgs/codemagic-ci-cd/discussions).
- [X] I confirm the software versions on my local computer match the ones in CI environment (e.g. Xcode version, Gradle version). See [here](https://docs.codemagic.io/troubleshooting/common-android-issues/#builds-work-locally-but-fail-on-codemagic)
- [X] I confirm I am able to do a fresh clone of the repository and build the project on my local computer.

### Which mobile framework are you using?

Flutter (Dart)

### Steps to reproduce

It seems that for some reason there are problems with `ruby` on machine associated with the app/build/account. The problem only appears on `mac_pro` machine. 
I tried:
1. Temporarily switching from `mac_pro` to `mac_mini_m1`. Everything worked with the same configuration. 
2. Selecting a different version of ruby: 
```bash
[  +78 ms] rbenv: pod: command not found

           The `pod' command exists in these Ruby versions:
             2.7.2
```
For some reason, it seems that `codemagic.yaml` does not respect the order of dependencies, i.e. installs `cocoapods` first and then switches version of `ruby`. Of course, the installation will be broken.

Can I somehow clear the cache for my `mac_pro` machine image? Or change the version of ruby but with `gem install cocoapods` running after it, not before? Thanks

### Expected results

The iOS build was successful

### Actual results

The iOS build failed with 
```bash
[   +7 ms] executing: pod --version
[ +404 ms] /Users/builder/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/activesupport-7.0.1/lib/active_support/xml_mini.rb:184:in `current_thread_backend': uninitialized constant ActiveSupport::XmlMini::IsolatedExecutionState (NameError)
           	from /Users/builder/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/activesupport-7.0.1/lib/active_support/xml_mini.rb:103:in `backend='
           	from /Users/builder/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/activesupport-7.0.1/lib/active_support/xml_mini.rb:201:in `<module:ActiveSupport>'
           	from /Users/builder/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/activesupport-7.0.1/lib/active_support/xml_mini.rb:11:in `<top (required)>'
           	from /Users/builder/.rbenv/versions/2.7.2/lib/ruby/site_ruby/2.7.0/rubygems/core_ext/kernel_require.rb:85:in `require'
           	from /Users/builder/.rbenv/versions/2.7.2/lib/ruby/site_ruby/2.7.0/rubygems/core_ext/kernel_require.rb:85:in `require'
           	from /Users/builder/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/activesupport-7.0.1/lib/active_support/core_ext/array/conversions.rb:3:in `<top (required)>'
           	from /Users/builder/.rbenv/versions/2.7.2/lib/ruby/site_ruby/2.7.0/rubygems/core_ext/kernel_require.rb:85:in `require'
           	from /Users/builder/.rbenv/versions/2.7.2/lib/ruby/site_ruby/2.7.0/rubygems/core_ext/kernel_require.rb:85:in `require'
           	from /Users/builder/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/cocoapods-1.12.1/lib/cocoapods.rb:9:in `<top (required)>'
           	from /Users/builder/.rbenv/versions/2.7.2/lib/ruby/site_ruby/2.7.0/rubygems/core_ext/kernel_require.rb:85:in `require'
           	from /Users/builder/.rbenv/versions/2.7.2/lib/ruby/site_ruby/2.7.0/rubygems/core_ext/kernel_require.rb:85:in `require'
           	from /Users/builder/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/cocoapods-1.12.1/bin/pod:36:in `<top (required)>'
           	from /Users/builder/.rbenv/versions/2.7.2/bin/pod:25:in `load'
           	from /Users/builder/.rbenv/versions/2.7.2/bin/pod:25:in `<main>'
[   +1 ms] Warning: CocoaPods is installed but broken. Skipping pod install.
             You appear to have CocoaPods installed but it is not working.
             This can happen if the version of Ruby that CocoaPods was installed with is different from the one being used to invoke it.
             This can usually be fixed by re-installing CocoaPods.
           To re-install see https://guides.cocoapods.org/using/getting-started.html#installation for instructions.

[  +22 ms] "flutter ios" took 7,510ms.
[   +4 ms] CocoaPods not installed or not in valid state.
[        ] 
           #0      throwToolExit (package:flutter_tools/src/base/common.dart:10:3)
 ```

### Build id (optional)

64a2c0830e427296dd48a171

**Comment by vysotsky** [comment]
The same configuration but `mac_mini_m1` - 64a2b2ab0e427296dd4877b7

**Comment by himesh-cm** [comment]
Hi @vysotsky, Checking the machine [specs](https://docs.codemagic.io/specs/versions-macos/), M1's uses `ruby 3.0.4p208` while the `mac pro` machines use `ruby 2.7.2p137`.
You can change the ruby version by specifying it in `yaml` for mac pro instances. Please try and let us know how it goes.
```
environment:
  ruby: 3.0.5
```

**Reply by himesh-cm** [reply]
Please note that your team is on a paid plan. As a paid member, use our chat widget for fast and direct support. It's the quickest way to get help.

**Reply by himesh-cm** [reply]
If you have tried it as per my above suggestion, please share the build id so I can take a look at it.

**Reply by vysotsky** [reply]
Hi @himesh-cm 
Thanks for checking it out. As I said, I tried a different version of Ruby (3.0.12). The thing is that `gem install cocoapods` is executed before `rbenv` selects the Ruby version. As a result, the gem installation will be broken. I assume that it might be some problem on your side, but it's strange that it only affected me and my team. Please check the build `64a2c67e259e925a8d480d8d` for more details.

Is there any way to request an updated for the `mac_pro` machines to use some Ruby 3.0.+ as the default Ruby? Because even if the version change works, according to [your documentation](https://docs.codemagic.io/yaml-basic-configuration/yaml-getting-started/#environment), manually changing the version will significantly increase the time of `Preparing build machine step`.
Also, you mentioned that I had to use a support chat widget. Unfortunately, we don't have it on different browsers with Adblock enabled and disabled. That's why I'm writing to you here.
Thank you

**Reply by himesh-cm** [reply]
Hi @vysotsky, Let me check that; also, is there any specific reason for opting for intel machines rather than M1?

**Reply by vysotsky** [reply]
@himesh-cm thank you
Yes. Unfortunately, 8 GB of RAM is not enough for us, and Gradle crashes with an OOM exception. We will try to reduce the memory usage later, but it's not a priority right now.

**Reply by himesh-cm** [reply]
Hi @vysotsky,
please setup this script, it should fix the error `rbenv: pod: command not found`

```
scripts:
      - name: install ruby 3.0.5
         script: |
           ruby --version
           rbenv install 3.0.5
           export RBENV_VERSION=3.0.5
           ruby --version
           rbenv global 3.0.5
```

Regarding the default version, i will pass the info to the devs.

**Reply by vysotsky** [reply]
@himesh-cm 
Unfortunately, the possible workaround you've provided didn't help:
```] Warning: CocoaPods is installed but broken. Skipping pod install.
             You appear to have CocoaPods installed but it is not working.
             This can happen if the version of Ruby that CocoaPods was installed with is different from the one being used to invoke it.
             This can usually be fixed by re-installing CocoaPods.
           To re-install see https://guides.cocoapods.org/using/getting-started.html#installation for instructions.

[  +22 ms] "flutter ios" took 7,240ms.
[   +4 ms] CocoaPods not installed or not in valid state.
[        ] 
```

The build id is `64a6aba9c73468f639fc56e9`

**Reply by himesh-cm** [reply]
Hi @vysotsky, on checking, I can see that your latest build was successful; what steps you took to resolve the ruby version issue?

**Comment by himesh-cm** [comment]
Hi @vysotsky, I'm tentatively closing the discussion since we haven't heard back from you. You can open the discussion again by replying to this message or opening a new discussion.
